Datagen

delhomer opened the pull request:
This PR introduces a new module dedicated to the dataset creation. Before to train a model on a given dataset, we choose to write the dataset on the file system, for further reusing. The following command is then expected, as an example:
```
python deeposlandia/datagen.py -d mapillary -s 224 -a -t 200 -v 20 -T 50
```
In this example, the program will generate 200 training images, 20 validation images and 50 testing images for the Mapillary dataset, and consequently produce 224*224 preprocessed images. As a remark, testing dataset generation implies only image generation, without labels.

Some additional tests has been developed as well to ensure the effectiveness of dataset creation, populating and loading operations (see `tests/test_dataset.py`). Concurrently a new data sample have been added to the `tests` repo, so as to accomplish the new tests.

From now, we do consider that dataset creation is not in the scope of the training program. Consequently, instead of creating the dataset before the training process if the former does not exist on the file system, the latter will just crash by telling there is no dataset (and eventually, by advising to launch the corresponding dataset creation command).

delhomer committed:
dataset: drop useless getter methods from Dataset class

get_label() and get_image() are not used, they can be deleted.

delhomer committed:
dataset: fix get_nb_labels method name

delhomer committed:
dataset: modify get_nb_labels to manage case where non-evaluated labels are included

Before this commit, get_nb_labels method returned the number of
evaluated labels, by letting aside non-evaluated ones. From now, a
parameter can be passed to this function to consider all labels, if
needed.

delhomer committed:
tests: add new tests dedicated to dataset creation, populating and loading

This commit introduces a range of new test functions related to
datasets. A bunch of test images are added to the `./tests/`
repository as well, in order to succeed in these new tests. They are
stored in a `validation/` subdirectory.

delhomer committed:
dataset: Add a new module to manage dataset creation

This commit allows to create easily new image datasets, regarding a
dataset type (`mapillary` or `shapes`), an image size and image
quantities. A particular care is provided to datapaths, so as to
respect some internal path conventions. When the dataset is created
(or if it is already on the file system), the label popularity inside
the training set is printed onto the console.

delhomer committed:
dataset: generate testing datasets

The previous commits allow to generate and/or load training and
validation images, however testing images were missing. This new
commit solves the problem by considering a testing dataset during
datagen.py execution.

delhomer committed:
generator: Address the testing case when creating a data generator

In the testing case, we do not need labels (or they are just
unavailable). The data generator must take this fact into account by
including only images into the returned iterators.

garaud commented on deeposlandia/datagen.py:
```diff
@@ -0,0 +1,118 @@
+
+
+if __name__=='__main__':
+    """Main method: generate a dataset from some instance specifications
```
I think the string here is not necessary. This is not a docstring. If we want to keep it, we can turn this content into comments with `#`.

delhomer replied:
The line has been dropped.

garaud commented on deeposlandia/datagen.py:
```diff
@@ -0,0 +1,118 @@
+"""Main method to generate new datasets
+"""
```
Maybe you can add an usage example of this script, e.g.

```to generate data about the xxx dataset, do

python datagen.py --args ...
```

delhomer replied:
Two examples have been added to the module docstring.

garaud commented on tests/test_dataset.py:
```diff
@@ -0,0 +1,111 @@
+"""
+
+import json
+import numpy as np
```
numpy is imported but not used.

delhomer replied:
Import removed.

garaud commented on tests/test_dataset.py:
```diff
@@ -0,0 +1,111 @@
+    assert d.get_nb_labels(see_all=True) == NB_LABELS
+    assert d.get_nb_images() == 0
+
+def test_mapillary_dataset_population():
```
When you test the creation of several image files from a specific dataset, you modify the working directory. See the `git status` after a `pytest tests/test_dataset.py`.

I think, we should have:

* a directory to input data (JSON, some images, etc.)
* a directory to output data which has to be cleaned after each unit test. I think we can add some pytest fixtures to create output dir on the fly, generate files, check some stuff, and remove the output dir.

delhomer replied:
Ok, I guess that for now, the output of the corresponding tests has been committed. You think it could be better to create the image files and delete it on the fly, without committing them?

garaud replied:
I previously mentioned the [pytest fixtures](https://docs.pytest.org/en/latest/fixture.html). I think it's a proper and cleaner way to handle temporary directories. See the following example https://docs.pytest.org/en/latest/tmpdir.html

delhomer committed:
dataset: improve datagen.py module docstring by adding some examples

delhomer committed:
tests: remove a useless numpy import in test_dataset.py

garaud commented:
I think we cannot have commited test files such as JSON and images which
will be modified every time we launch the test.

And I think it's better to create and clean-up for each test.
Raphael Delhome writes:

> delhomer commented on this pull request.
>
>
>
>> +    assert d.get_nb_labels() == 1
> +
> +def test_mapillary_dataset_creation():
> +    """Create a Mapillary dataset from a configuration file
> +    """
> +    IMAGE_SIZE = 224
> +    config_filename = "tests/data/mapillary/config_aggregate.json"
> +    with open(config_filename) as fobj:
> +        config = json.load(fobj)
> +    NB_LABELS = len(config["labels"])
> +    d = MapillaryDataset(IMAGE_SIZE, config_filename)
> +    assert d.image_size == IMAGE_SIZE
> +    assert d.get_nb_labels(see_all=True) == NB_LABELS
> +    assert d.get_nb_images() == 0
> +
> +def test_mapillary_dataset_population():
>
> Ok, I guess that for now, the output of the corresponding tests has been committed. You think it could be better to create the image files and delete it on the fly, without committing them?


-- 
Damien Garaud
Data Scientist – Developer, Oslandia
http://oslandia.com

delhomer committed:
tests: reorganized dataset samples in ./tests repo

Validation datasets are used only when testing dataset population
process. It is then quite useless to commit them and keep them on the
git repo. They are deleted, and corresponding raw Mapillary images are
moved to a dedicated repository in
./tests/data/mapillary/input_dataset_population/.

delhomer committed:
tests: create and delete dataset images on-the-fly when testing dataset population

With this commit, the images generated on the file system by dataset
population tests are dropped at the end of them. `os.removedirs` does
not the job, consequently `shutil` library is imported, and the
temporary image repository is dropped with `shutil.rmtree`.

delhomer commented:
Some changes has been brought with last commits:
- a training data sample is kept on the git repo, it is used for a wide range of tests (regarding datasets, generators and even model training).
- Mapillary raw data sample for dataset population testing is kept as well in a dedicated repository (`tests/data/mapillary/input_dataset_population` instead of `tests/data/mapillary/validation/raw/`)
- images that were created during dataset population tests are stored into `tests/data/dataset/tmp/` during the test, and deleted on-the-fly at the end of the tests

We could have replicated the data file tree (see issue #23 ), however as you said, files that are generated during tests are temporary per nature, most of the time. I hope it is OK with these changes.

garaud commented on deeposlandia/datagen.py:
````diff
@@ -0,0 +1,129 @@
+
+- generate 64*64 pixel images from Shapes dataset, 10000 images in the training set, 100 in the
+validation set, 1000 in the testing set:
+```
````
the docstring python syntax is often written in ReST, [Restructured
Text](http://docutils.sourceforge.net/rst.html) instead of markdow.

See http://docutils.sourceforge.net/docs/user/rst/quickstart.html

The ReST format
to write snippet of code can be:

```
The following snippet::

    import sys
    sys.prefix
```

with two `:` and at least four indentation whitespaces.

And if you want to inline it, you have to use two ticks such as:

```
you can call the function ``foobar``.
```

delhomer replied:
Ok, modified.

garaud commented:
despite the conflic with the module `generator.py` as specified by Github, it's good for me.
